File: pico_zero_interface/zero/spi_master.py
"""
SPI Master Interface for Pico <-> Zero communication.

Implements the Zero (master) side of the protocol described in
pico_zero_interface/README.md.

Uses spidev for SPI transfers and gpiod for IRQ/READY GPIO handling.
"""

import datetime
import struct
import time

import gpiod
import spidev
from gpiod.line import Bias, Direction, Edge

# --- Protocol constants (must match Pico side) ---

SPI_CMD_WRITE = 0x01
SPI_CMD_REQUEST = 0x02
SPI_CMD_READ = 0x03

READ_SIZE = 1503        # 3-byte header + 1500-byte payload
MAX_PAYLOAD = 1500

# --- Pin assignments (BCM numbering) ---

GPIO_CHIP = "/dev/gpiochip0"
PIN_IRQ = 25            # Input, active low ("Pico has data")
PIN_READY = 24          # Input, active low ("TX DMA loaded, safe to READ")

# --- SPI settings ---

SPI_BUS = 0
SPI_DEVICE = 0          # CE0
SPI_SPEED_HZ = 1_000_000
SPI_MODE = 3            # CPOL=1, CPHA=1 (PL022 slave requires Mode 1 or 3 for multi-byte)


class SpiMaster:
    """Zero-side SPI master for the Pico <-> Zero protocol."""

    def __init__(self, spi_speed_hz=SPI_SPEED_HZ):
        self._spi = spidev.SpiDev()
        self._spi.open(SPI_BUS, SPI_DEVICE)
        self._spi.mode = SPI_MODE
        self._spi.max_speed_hz = spi_speed_hz

        # Track last-known buffer space on Pico (in 64-byte units).
        # Start pessimistic -- do a REQUEST/READ to get the real value.
        self.buf = 0

        # gpiod line request for IRQ and READY
        self._gpio = gpiod.request_lines(
            GPIO_CHIP,
            consumer="spi_master",
            config={
                PIN_IRQ: gpiod.LineSettings(
                    direction=Direction.INPUT,
                    bias=Bias.PULL_UP,
                    edge_detection=Edge.FALLING,
                ),
                PIN_READY: gpiod.LineSettings(
                    direction=Direction.INPUT,
                    bias=Bias.PULL_UP,
                ),
            },
        )

    def close(self):
        """Release SPI and GPIO resources."""
        self._spi.close()
        self._gpio.release()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    # --- Low-level helpers ---

    def _irq_is_asserted(self):
        """Check if IRQ is currently low (Pico has data)."""
        return self._gpio.get_value(PIN_IRQ) == gpiod.line.Value.INACTIVE

    def _wait_ready(self, timeout_s=1.0):
        """Wait for READY to go low (Pico has loaded TX DMA).

        Returns True if READY asserted, False on timeout.
        """
        deadline = time.monotonic() + timeout_s
        while time.monotonic() < deadline:
            if self._gpio.get_value(PIN_READY) == gpiod.line.Value.INACTIVE:
                return True
            time.sleep(0.0001)  # 100us poll interval
        return False

    def _wait_ready_deasserted(self, timeout_s=1.0):
        """Wait for READY to go high after a READ completes.

        Per protocol: Zero must wait for READY high before new transactions.
        Returns True if READY deasserted, False on timeout.
        """
        deadline = time.monotonic() + timeout_s
        while time.monotonic() < deadline:
            if self._gpio.get_value(PIN_READY) == gpiod.line.Value.ACTIVE:
                return True
            time.sleep(0.0001)
        return False

    # --- Protocol operations ---

    def write(self, payload):
        """Send a WRITE transaction to the Pico.

        Args:
            payload: bytes or bytearray, up to MAX_PAYLOAD bytes.

        Returns:
            True if sent, False if payload too large or insufficient buffer.
        """
        if len(payload) > MAX_PAYLOAD:
            return False

        # Check local buffer estimate
        buf_needed = (len(payload) + 63) // 64  # Round up to 64-byte units
        if buf_needed > self.buf:
            return False

        header = struct.pack(">BH", SPI_CMD_WRITE, len(payload))
        self._spi.xfer2(list(header) + list(payload))

        # Decrement local buffer estimate
        self.buf -= buf_needed
        if self.buf < 0:
            self.buf = 0

        return True

    def request_and_read(self, timeout_s=1.0):
        """Send REQUEST, wait for READY, then send READ.

        Returns:
            (payload_bytes, buf_value) on success, or None on timeout/error.
        """
        # Step 1: Send REQUEST (single byte)
        self._spi.xfer2([SPI_CMD_REQUEST])

        # Step 2: Wait for READY
        if not self._wait_ready(timeout_s):
            return None

        # Step 3: Send READ (READ_SIZE bytes)
        tx_data = [SPI_CMD_READ] + [0x00] * (READ_SIZE - 1)
        rx_data = self._spi.xfer2(tx_data)

        # Step 4: Wait for READY to deassert before allowing new transactions
        self._wait_ready_deasserted(timeout_s=0.1)

        # Parse response: [LEN_HI] [LEN_LO] [BUF] [payload...]
        if len(rx_data) < 3:
            return None

        payload_len = (rx_data[0] << 8) | rx_data[1]
        self.buf = rx_data[2]

        if payload_len > MAX_PAYLOAD:
            payload_len = MAX_PAYLOAD  # Defensive

        payload = bytes(rx_data[3:3 + payload_len])
        return payload, self.buf

    def wait_for_irq(self, timeout_s=None):
        """Wait for the Pico to assert IRQ (falling edge).

        Args:
            timeout_s: Timeout in seconds, or None to block indefinitely.

        Returns:
            True if IRQ detected, False on timeout.
        """
        # Check if already asserted
        if self._irq_is_asserted():
            return True

        # Wait for falling edge event
        timeout_ns = None
        if timeout_s is not None:
            timeout_ns = int(timeout_s * 1e9)
            if timeout_ns <= 0:
                return False

        # gpiod v2: wait_edge_events takes timeout as timedelta or seconds
        if self._gpio.wait_edge_events(
            datetime.timedelta(seconds=timeout_s) if timeout_s else None
        ):
            # Consume the event
            self._gpio.read_edge_events()
            return True
        return False

File: shein/src/spi_master.rs
pub const MAX_PAYLOAD: usize = 1542; // 257*6: room for 6 max-size TLV packets
pub const NUM_DEVICES: usize = 8;

// ── Linux (real hardware) ───────────────────────────────────────────────────

#[cfg(target_os = "linux")]
mod hw {
    use std::io::Write;
    use std::time::{Duration, Instant};

    use anyhow::{Context, Result};
    use gpiocdev::line::{Bias, EdgeDetection, Value};
    use gpiocdev::Request;
    use spidev::{Spidev, SpidevOptions, SpidevTransfer, SpiModeFlags};

    const SPI_CMD_WRITE: u8 = 0x01;
    const SPI_CMD_REQUEST: u8 = 0x02;
    const SPI_CMD_READ: u8 = 0x03;

    const READ_SIZE: usize = super::MAX_PAYLOAD + 10; // 8 buf + 2 len + payload

    const GPIO_CHIP: &str = "/dev/gpiochip0";
    const PIN_IRQ: u32 = 25;
    const PIN_READY: u32 = 24;

    const SPI_DEVICE: &str = "/dev/spidev0.0";
    const SPI_SPEED_HZ: u32 = 8_000_000;

    pub struct IrqWatcher {
        req: Request,
    }

    impl IrqWatcher {
        pub fn new() -> Result<Self> {
            let req = Request::builder()
                .on_chip(GPIO_CHIP)
                .with_line(PIN_IRQ)
                .as_input()
                .with_bias(Bias::PullUp)
                .with_edge_detection(EdgeDetection::FallingEdge)
                .with_consumer("shein-irq")
                .request()
                .context("Failed to request IRQ GPIO")?;
            Ok(Self { req })
        }

        pub fn is_asserted(&self) -> Result<bool> {
            let val = self.req.value(PIN_IRQ)?;
            Ok(val == Value::Inactive)
        }

        pub fn wait_edge(&self, timeout: Duration) -> Result<bool> {
            Ok(self.req.wait_edge_event(timeout)?)
        }

        pub fn consume_edge(&self) -> Result<()> {
            self.req.read_edge_event()?;
            Ok(())
        }
    }

    pub struct SpiMaster {
        spi: Spidev,
        ready: Request,
        pub buf: [u8; super::NUM_DEVICES],
    }

    impl SpiMaster {
        pub fn new() -> Result<Self> {
            let mut spi = Spidev::open(SPI_DEVICE).context("Failed to open SPI device")?;
            let options = SpidevOptions::new()
                .bits_per_word(8)
                .max_speed_hz(SPI_SPEED_HZ)
                .mode(SpiModeFlags::SPI_MODE_3)
                .build();
            spi.configure(&options).context("Failed to configure SPI")?;

            let ready = Request::builder()
                .on_chip(GPIO_CHIP)
                .with_line(PIN_READY)
                .as_input()
                .with_bias(Bias::PullUp)
                .with_consumer("shein-ready")
                .request()
                .context("Failed to request READY GPIO")?;

            Ok(Self { spi, ready, buf: [0u8; super::NUM_DEVICES] })
        }

        fn wait_ready(&self, timeout: Duration) -> Result<bool> {
            let deadline = Instant::now() + timeout;
            while Instant::now() < deadline {
                if self.ready.value(PIN_READY)? == Value::Inactive {
                    return Ok(true);
                }
                std::thread::sleep(Duration::from_micros(100));
            }
            Ok(false)
        }

        fn wait_ready_deasserted(&self, timeout: Duration) -> Result<bool> {
            let deadline = Instant::now() + timeout;
            while Instant::now() < deadline {
                if self.ready.value(PIN_READY)? == Value::Active {
                    return Ok(true);
                }
                std::thread::sleep(Duration::from_micros(100));
            }
            Ok(false)
        }

        pub fn write(&mut self, payload: &[u8]) -> Result<bool> {
            if payload.len() > super::MAX_PAYLOAD {
                return Ok(false);
            }

            let len = payload.len() as u16;
            let mut tx = Vec::with_capacity(3 + payload.len());
            tx.push(SPI_CMD_WRITE);
            tx.push((len >> 8) as u8);
            tx.push((len & 0xFF) as u8);
            tx.extend_from_slice(payload);

            self.spi
                .write_all(&tx)
                .context("SPI WRITE transfer failed")?;

            Ok(true)
        }

        pub fn request_and_read(&mut self, timeout: Duration) -> Result<Option<(Vec<u8>, [u8; super::NUM_DEVICES])>> {
            self.spi
                .write_all(&[SPI_CMD_REQUEST])
                .context("SPI REQUEST transfer failed")?;

            if !self.wait_ready(timeout)? {
                return Ok(None);
            }

            let mut tx_buf = vec![0u8; READ_SIZE];
            tx_buf[0] = SPI_CMD_READ;
            let mut rx_buf = vec![0u8; READ_SIZE];

            let mut transfer = SpidevTransfer::read_write(&tx_buf, &mut rx_buf);
            self.spi
                .transfer(&mut transfer)
                .context("SPI READ transfer failed")?;

            let _ = self.wait_ready_deasserted(Duration::from_millis(100));

            // Bytes 0..8: per-device buffer estimates (in 16-byte units)
            self.buf.copy_from_slice(&rx_buf[0..super::NUM_DEVICES]);

            // Bytes 8..10: payload length (big-endian)
            let payload_len = ((rx_buf[8] as usize) << 8) | (rx_buf[9] as usize);
            let payload_len = payload_len.min(super::MAX_PAYLOAD);
            let payload = rx_buf[10..10 + payload_len].to_vec();

            Ok(Some((payload, self.buf)))
        }
    }
}

// ── Stub (non-Linux, for development/testing) ───────────────────────────────

#[cfg(not(target_os = "linux"))]
mod hw {
    use std::time::Duration;
    use anyhow::Result;

    pub struct IrqWatcher;

    impl IrqWatcher {
        pub fn new() -> Result<Self> {
            Ok(Self)
        }

        pub fn is_asserted(&self) -> Result<bool> {
            Ok(false)
        }

        pub fn wait_edge(&self, timeout: Duration) -> Result<bool> {
            std::thread::sleep(timeout);
            Ok(false)
        }

        pub fn consume_edge(&self) -> Result<()> {
            Ok(())
        }
    }

    pub struct SpiMaster {
        pub buf: [u8; super::NUM_DEVICES],
    }

    impl SpiMaster {
        pub fn new() -> Result<Self> {
            Ok(Self { buf: [255u8; super::NUM_DEVICES] })
        }

        pub fn write(&mut self, payload: &[u8]) -> Result<bool> {
            if payload.len() > super::MAX_PAYLOAD {
                return Ok(false);
            }
            Ok(true)
        }

        pub fn request_and_read(&mut self, _timeout: Duration) -> Result<Option<(Vec<u8>, [u8; super::NUM_DEVICES])>> {
            self.buf = [255u8; super::NUM_DEVICES];
            Ok(Some((Vec::new(), self.buf)))
        }
    }
}

pub use hw::{IrqWatcher, SpiMaster};
